### ISSUE TITLE
Satellites crashing with Unhandled IllegalStateException / need of liveness probes for piraeus-op-ns-node pods

### ISSUE THREAD

**Issue by Boca13** [piraeusdatastore/piraeus-operator#176]
Hello,

First of all, I'd like to say that after a lot of research and testing several solutions, I've found Piraeus/Linstor to be one of the best performant storage solutions (at least for my workloads). Thank you for creating maintaining the Piraeus operator, which I think is the best way to use Linstor in Kubernetes.

I've decided to use Piraeus in my cluster mainly because its great read performance, but I'm having several issues with stability as I add more volumes. It would be great if I could fix these problems and trust Piraeus with important workloads.

I'm currently running Piraeus 1.4.0 installed with Helm in a Kubernetes cluster with 4 worker nodes/satellites.

## Satellites crashing
My satellites are crashing quite often, and the logs always end like this:
```
16:22:19.177 [MainWorkerPool-1] ERROR LINSTOR/Satellite - SYSTEM - Failed to execute blkid [Report number 609170B7-60BF5-000111]

16:22:20.074 [SSLNetComService] ERROR LINSTOR/Satellite - SYSTEM - Unhandled IllegalStateException [Report number 609170B7-60BF5-000113]
```
After that message, it seems like the satellites are completely crashed and they start appearing as OFFLINE in the controller. I don't know how to retrieve more information about those reports. Doing a `linstor error-reports show` in the controller, doesn't provide any information.

The only way around this is to delete the `piraeus-op-ns-node-xxxxx` pod, wait until it's Running and then manually issue a `linstor node restore`. Often, these pods get stuck in Terminating state.

These crashes seem more frequent as the number of Online satellites increases. I can't manage to have all four satellites online at the same time for longer than a few seconds. Many volumes have become impossible to mount. How do you think I could investigate this problem?

## Latency
I've been searching in the documentation and the Linbit-community Slack but I couldn't come to any conclusions about this: How much does latency affect Piraeus and Linstor? In my cluster, the nodes are cloud servers and one dedicated node. The ping between these zones is ~3ms (up to 4ms). I'm planning on moving all the workers to the same zone, where ping would be <1ms (avg 0.6ms). Is Piraeus so sensitive to latency (other than for write operations)?

## Probes
I've seen that the `piraeus-op-ns-node-xxxxx` pods have a readinessProbe, but I think a livenessProbe might be useful to restart the pod when it becomes unresponsive. However, the readiness probe fails to detect this state, since the TCP connection actually happens (returning a "P" apparently). Any hints on how to automatically detect these situations?

![image](https://user-images.githubusercontent.com/1238543/117044451-bef7f280-ad0e-11eb-91f0-6fab7df1c18a.png)

Thank you!

## Edit: stuck blkid processes in host
In one of the nodes (dedicated-1), system load seems too high (~20 for a 12-thread machine) but CPU usage is not. Running `ps -e v` outputs 15 processes like this: 
```
1813203 ?        D      0:00      0    81  2678   836  0.0 blkid -p -s TYPE -s PTTYPE -o export /dev/drbd1022
```

**Comment by WanzenBug** [comment]
Hi! Thanks for the detailed write-up and investigation.

I'd hazard a guess: Something is blocking blkid and tools like it from accessing data on drbd devices. This can have a number of reasons:
* Multipathd opens DRBD devices if not configured on some distributions (I know it's a problem on ubuntu 20.04, not sure on others). If you are running multipath, you need to block it from accessing DRBD devices (taken from a `/etc/multipath/conf.d/drbd.conf`):
```
blacklist {
        devnode "^drbd[0-9]+"
}
```
Also note: very recent versions of mulitpathd already ship their own blacklist that should exclude DRBD devices.

* LVM tools also need to be told to not scan DRBD devices. An example taken from `/etc/lvm.conf`:
```
devices {
	# Disable scanning drbd devices
	global_filter = [ "r|^/dev/drbd.*|" ]
}
```

* Some host monitoring agents (I think Zabbix is one such example) also keep the device open.

So I guess I'd check out if you are one of these cases above and try to configure the node to leave DRBD devices alone ;-)

**Comment by Boca13** [comment]
Hi Moritz, thanks for your response and your tips. I tried what you mentioned, but I'm still having trouble:
- Multipathd: I checked and the settings are correct in the hosts.
- LVM config: I tried that before posting the issue. After adding the global filter, lvdisplay started working. However, I'm still having problems with blkid
- Monitoring: The only monitoring tool that I'm using is Prometheus Node Exporter.

Other things I've tried:
- Kernel module injection: I tried switching from host-install to letting the kernel-module-injector do the job. For some reason, the DRBD version installed with this method was 8, and it didn't seem able to load the data created with DRBD v9. I installed DRBD again with apt: `apt install drbd-utils drbd-dkms` (the hosts run Ubuntu 20.04 and I added the Linbit repository `ppa:linbit/linbit-drbd9-stack`).
- I checked that `/sys/module/drbd/parameters/usermode_helper` is disabled in all hosts.

About blkid:
- I scaled down to zero all the workloads using Piraeus volumes and then I tried rebooting one of the hosts nodes. blkid worked properly right after, so I think the problem is related to Piraeus.
- blkid kept working even after the satellite pod started and connected to the controller.
- When I scaled a deployment back up, the new pod was created successfully and started running normally. However, blkid started freezing again in the host and the usual problem happened again: satellites crash and get disconnected from the controller after a few `Failed to execute blkid` messages. I checked the error report but it doesn't say anything interesting, just that the error happens because of a timeout executing the external command blkid.

One question: I've seen that the workload pods often get deleted after a satellite failure. They might be working fine but suddenly all pods using Piraeus in the failing node will be replaced by `Pending` pods. What component of Piraeus is responsible for this behavior? It adds extra instability.

I hope we can figure it out, because Piraeus is great. Thank you again.

**Comment by WanzenBug** [comment]
> One question: I've seen that the workload pods often get deleted after a satellite failure. They might be working fine but suddenly all pods using Piraeus in the failing node will be replaced by Pending pods. What component of Piraeus is responsible for this behavior? It adds extra instability.

That is most likely Stork, if you have it enabled. At some point it gained some rudimentary "volume health" feature, so I guess you experience it in action. You can just scale down both stork and the stork-scheduler deployment or delete them altogether if you are currently not using them.

Regarding the blkid issue: do you see any stack traces in the kernel log, perhaps something like "task xxx hung for 120 seconds"? If so, could you post them, that could help pinpoint the issue.

**Comment by Boca13** [comment]
Hello again Moritz,

I think the problem got fixed.

I couldn't find any relevant messages in the kernel log, so I ran out of ideas. However, when I saw the new Piraeus release (1.5) I decided to give it a try and see. The uprade didn't work (I'll propably open a new issue for that), but I tried changing the kernel-module-injector image to 9.0.29 and that seems to have fixed. I'll detail all my steps in case it helps you or other users that might face similar problems:
1. Even though latency was low and shouldn't have been an issue, I consolidated all workers that run Piraeus in the same region to see if it had any effects. I don't think it helped.
2. I increased and improved the resources assigned to Piraeus' etcd cluster, which I think made everything run smoother and the initial load time for the controller improved a lot.
3. Updated to DRBD 9.0.29: to do so, I changed the `kernelModuleInjectionImage` field in my `linstorsatelliteset` object. The containers still used DRBD 9.0.28 because it was already loaded in the kernel, so I made sure that there weren't any DRBD-related packages installed with apt and then restarted the hosts. This way, the injector init container compiled DRBD 9.0.29 and suddenly everything started working. blkid doesn't hang and the satellites never crash or get disconnected from the controller anymore. The fix could be related either to the new version or to the fact that I "reinstalled" DRBD.

I'll close the issue for now. Thank you for your help, @WanzenBug .
